be/dtm0_log.c

```diff
@@ -0,0 +1,631 @@
+{
+	return _0C(log != NULL) &&
+	       _0C(log->dl_cs != NULL); // &&
+	/* TODO: Add an invariant check against the volatile part */
+	       /* _0C(lrec_tlist_invariant(log->dl_tlist)); */
+}
+
+static bool m0_dtm0_log_rec__invariant(const struct m0_dtm0_log_rec *rec)
+{
+	return _0C(rec != NULL) &&
+	       _0C(m0_dtm0_tx_desc__invariant(&rec->dlr_txd)); // &&
+	       /* _0C(m0_tlink_invariant(&lrec_tl, rec)); */
+}
+
+M0_INTERNAL int m0_be_dtm0_log_init(struct m0_be_dtm0_log **out,
+                                    struct m0_dtm0_clk_src *cs,
+                                    bool                    is_plog)
+{
+	struct m0_be_dtm0_log *log;
+
+	M0_PRE(out != NULL);
+	M0_PRE(cs != NULL);
+
+	if (!is_plog) {
+		M0_ALLOC_PTR(log);
+		if (log == NULL)
+			return M0_ERR(-ENOMEM);
+
+		M0_SET0(log);
+		M0_ALLOC_PTR(log->dl_tlist);
+		if (log->dl_tlist == NULL) {
+			m0_free(log);
+			return M0_ERR(-ENOMEM);
+		}
+		lrec_tlist_init(log->dl_tlist);
+		*out = log;
+	} else {
+		log = *out;
+	}
+
+	m0_mutex_init(&log->dl_lock);
+	log->dl_is_persistent = is_plog;
+	log->dl_cs = cs;
+	return 0;
+}
+
+M0_INTERNAL void m0_be_dtm0_log_fini(struct m0_be_dtm0_log **log,
+                                     bool                    isvstore)
+{
+	struct m0_be_dtm0_log *plog = *log;
+
+	M0_PRE(m0_be_dtm0_log__invariant(plog));
+	m0_mutex_fini(&plog->dl_lock);
+	lrec_tlist_fini(plog->dl_tlist);
+	plog->dl_cs = NULL;
+
+	if (isvstore) {
+		m0_free(plog->dl_tlist);
+		m0_free(plog);
+		*log = NULL;
+	}
+}
+
+M0_INTERNAL void dtm0_be_alloc_tx_desc_credit(struct m0_dtm0_tx_desc *txd,
+					      struct m0_be_seg *seg,
+					      struct m0_be_tx_credit *accum)
+{
+	M0_BE_ALLOC_CREDIT_PTR(txd, seg, accum);
+	M0_BE_ALLOC_CREDIT_ARR(txd->dtd_ps.dtp_pa,
+			       txd->dtd_ps.dtp_nr, seg, accum);
+}
+
+M0_INTERNAL void dtm0_be_alloc_log_rec_credit(struct m0_dtm0_tx_desc *txd,
+					      struct m0_buf	     *payload,
+					      struct m0_be_seg	     *seg,
+					      struct m0_be_tx_credit *accum)
+{
+	struct m0_dtm0_log_rec *rec;
+
+	M0_BE_ALLOC_CREDIT_PTR(rec, seg, accum);
+	M0_BE_ALLOC_CREDIT_BUF(payload, seg, accum);
+	dtm0_be_alloc_tx_desc_credit(txd, seg, accum);
+	lrec_be_list_credit(M0_BLO_TLINK_CREATE, 1, accum);
+}
+
+M0_INTERNAL void dtm0_be_set_log_rec_credit(struct m0_dtm0_tx_desc *txd,
+					     struct m0_buf	    *payload,
+					     struct m0_be_seg	    *seg,
+					     struct m0_be_tx_credit *accum)
+{
+	struct m0_dtm0_log_rec *rec;
+
+	M0_BE_ALLOC_CREDIT_BUF(payload, seg, accum);
+	m0_be_tx_credit_add(accum, &M0_BE_TX_CREDIT_TYPE(*payload));
+	m0_be_tx_credit_add(accum, &M0_BE_TX_CREDIT_TYPE(*txd));
+	m0_be_tx_credit_add(accum,
+			    &M0_BE_TX_CREDIT(txd->dtd_ps.dtp_nr,
+					     sizeof(struct m0_dtm0_tx_pa *)));
+	m0_be_tx_credit_add(accum,
+			    &M0_BE_TX_CREDIT(txd->dtd_ps.dtp_nr,
+					     sizeof(struct m0_dtm0_tx_pa)));
+	m0_be_tx_credit_add(accum, &M0_BE_TX_CREDIT_TYPE(*rec));
```

Comment:
The variable "rec" is only used to calculate the size of this data structure.
But this seems confusing here, because it is a pointer, not initialized.

Reply:
Initialized the pointer to NULL. This should silence codacy.